packages/vite/src/node/config.ts

```diff
@@ -1571,200 +1555,74 @@ export async function loadConfigFromFile(
+        }),
       },
-    ],
-  })
-  const { text } = result.outputFiles[0]
-  return {
-    code: text,
-    dependencies: result.metafile ? Object.keys(result.metafile.inputs) : [],
-  }
-}
-
-interface NodeModuleWithCompile extends NodeModule {
-  _compile(code: string, filename: string): any
-}
-
-const _require = createRequire(import.meta.url)
-async function loadConfigFromBundledFile(
-  fileName: string,
-  bundledCode: string,
-  isESM: boolean,
-): Promise<UserConfigExport> {
-  // for esm, before we can register loaders without requiring users to run node
-  // with --experimental-loader themselves, we have to do a hack here:
-  // write it to disk, load it with native Node ESM, then delete the file.
-  if (isESM) {
-    const nodeModulesDir = findNearestNodeModules(path.dirname(fileName))
-    if (nodeModulesDir) {
-      await fsp.mkdir(path.resolve(nodeModulesDir, '.vite-temp/'), {
-        recursive: true,
-      })
+      hot: false,
+    },
+  )
+  await environment.init()
+  try {
+    const { default: userConfig } = (await environment.runner.import(
+      resolvedPath,
+    )) as {
+      default: UserConfigExport
     }
-    const hash = `timestamp-${Date.now()}-${Math.random().toString(16).slice(2)}`
-    const tempFileName = nodeModulesDir
-      ? path.resolve(
-          nodeModulesDir,
-          `.vite-temp/${path.basename(fileName)}.${hash}.mjs`,
+    const modules = [
+      ...environment.runner.evaluatedModules.fileToModulesMap.entries(),
+    ]
+    const dependencies = modules
+      .filter(([file, modules]) => {
+        const isExternal = [...modules].every(
+          (m) => !m.meta || 'externalize' in m.meta,
         )
-      : `${fileName}.${hash}.mjs`
-    await fsp.writeFile(tempFileName, bundledCode)
-    try {
-      return (await import(pathToFileURL(tempFileName).href)).default
-    } finally {
-      fs.unlink(tempFileName, () => {}) // Ignore errors
-    }
-  }
-  // for cjs, we can register a custom loader via `_require.extensions`
-  else {
-    const extension = path.extname(fileName)
-    // We don't use fsp.realpath() here because it has the same behaviour as
-    // fs.realpath.native. On some Windows systems, it returns uppercase volume
-    // letters (e.g. "C:\") while the Node.js loader uses lowercase volume letters.
-    // See https://github.com/vitejs/vite/issues/12923
-    const realFileName = await promisifiedRealpath(fileName)
-    const loaderExt = extension in _require.extensions ? extension : '.js'
-    const defaultLoader = _require.extensions[loaderExt]!
-    _require.extensions[loaderExt] = (module: NodeModule, filename: string) => {
-      if (filename === realFileName) {
-        ;(module as NodeModuleWithCompile)._compile(bundledCode, filename)
-      } else {
-        defaultLoader(module, filename)
-      }
+        return !isExternal && file !== resolvedPath
+      })
+      .map(([file]) => file)
+    return {
+      userConfig,
+      dependencies,
```

Comment:
Maybe here we can expose a `close` callback that can free up mem?

Reply:
For now, I am just always closing the environment. I think if people want to use import files repeatedly, they can construct their own module runner. Or we can also expose something like `createInlinedModuleRunner` that creates the environment all by itself